egs/wsj/s5/utils/lang/make_subword_lexicon_fst.py

```diff
@@ -0,0 +1,240 @@
+            current_state = next_state
+            next_state += 1
+
+        i = len(phones) - 1
+        phone = (phones[i] if i >= 0 else constant_variable.eps)
+        word = (word if i <= 0 else constant_variable.eps)
+        cost = (pron_cost if i <= 0 else 0.0)
+        print_arc(current_state, to_state, phone, word, cost)
+
+        return next_state
+
+    loop_state = 0
+    prefix_state = 1
+    next_state = 2
+
+    for (word, pron_prob, phones) in lexicon:
+        cost = -math.log(pron_prob)
+
+        if is_end(word,separator):
+            if position_dependent:
+                position = get_position(phones, constant_variable)
+                next_state = print_word_no_silence(word, cost, phones, position, next_state)
+            else:
+                next_state = print_word_no_silence(word, cost, phones, constant_variable.end, next_state)
+                next_state = print_word_no_silence(word, cost, phones, constant_variable.word, next_state)
+        else:
+            # the word end with separator which should be followed by subword
+            if position_dependent:
+                position = get_position(phones, constant_variable) 
+                next_state = print_word_no_silence(word, cost, phones, position, next_state)
+            else:
+                next_state = print_word_no_silence(word, cost, phones, constant_variable.begin, next_state)
+                next_state = print_word_no_silence(word, cost, phones, constant_variable.middle, next_state)
+
+    print("{state}\t{final_cost}".format(state=loop_state, final_cost=0.0))
+
+def write_fst_with_silence(lexicon, sil_phone, sil_prob, sil_disambig, 
+    position_dependent, separator, constant_variable):
+    """Writes the text format of L.fst to the standard output.  This version is for
+       when --sil-prob != 0.0, meaning there is optional silence
+     'lexicon' is a list of 3-tuples (subword, pron-prob, prons)
+         as returned by read_lexiconp().
+     'sil_prob', which is expected to be strictly between 0.0 and 1.0, is the
+         probability of silence
+     'sil_phone' is the silence phone, e.g. "SIL".
+     'sil_disambig' is either None, or the silence disambiguation symbol, e.g. "#5".
+     'position_dependent', which is True is the lexicion is position-dependent.
+     'separator' is the symbol we ues to indicate the postion of a subword in word.
+    """
+    def print_word_with_silence(word, pron_cost, phones, position, next_state):
+    
+        assert(len(phones)) > 0
+    
+        from_state = prefix_state
+        to_state = prefix_state
+        if position == constant_variable.begin:
+            from_state = loop_state
+        elif position == constant_variable.end: 
+            to_state = constant_variable.sil_nonsil
+        elif position == constant_variable.word:
+            from_state = loop_state
+            to_state = constant_variable.sil_nonsil
```

Comment:
constant_variable.sil_nonsil seems to be string valued but the other states seem to be numeric.  This is not a good design.  I see that later on you test the value, but it should always be clear what variables mean.  This is a bit too spaghetti-like.